### PR TITLE
Password policies

### PR DESCRIPTION
**Pull request by xmo-odoo**
Task 31122 sections 2 and 3:

* enforce minimum length on passwords
* add a "strength meter" on password field, which is red if below db-configured minimum length, yellow beyond, and green if the password clears policy recommendations from Shay[2016]: 2word16 or  3class12

TODO:

- [x] auth_signup (?) support
- [x] check & disable (if necessary) existing `@password` test
- [x] check `@password` formatValue bug

### CODE REVIEW

File: addons/auth_password_policy/models/res_config_settings.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
+from odoo import api, fields, models, _
+
+
+class ResConfigSettings(models.TransientModel):
+    _inherit = 'res.config.settings'
+
+    minlength = fields.Integer("Minimum Length", help="Minimum number of characters passwords must contain.")
```
Comment (Elkasitu) [inline addons/auth_password_policy/models/res_config_settings.py]: what do you think of having a maxlength as well?

Comment (jorenvo) [reply addons/auth_password_policy/models/res_config_settings.py]: FWIW, personally I really dislike a maximum length on passwords. I dislike it so much I don't think the user should even be able to configure it :D. I think historically there were some technical limitations that somehow forced maximum password lengths (see e.g. the second to last paragraph [here](https://security.stackexchange.com/a/33471)). But we don't have those in Odoo. Additionally, these days a lot of people hopefully use password managers (the only way to do passwords right IMHO) and I think they should be allowed to generate really long passwords.

Comment (nhomar) [reply addons/auth_password_policy/models/res_config_settings.py]: name it min_length (I know stupid topic but it is clearer).

**Comment by nhomar** [comment]
Sorry people, but what a weak feature:

Did not you know this one?.

https://github.com/OCA/server-tools/tree/10.0/password_security

This is a real password policy module ;-)

and IMHO this should be a base module to avoid maintain several modules for such basic feature that should be a base one.

Well at the end an opinion, but I honestly do not know why this took too long to be an official feature.

BTW: migrated to 11.0 [here](https://github.com/Vauxoo/server-tools/tree/11.0/password_security) (no time yet to make a proper PR to OCA but I think convert that feature into an official one (fixing what ever you do not like) should be nice.

**Comment by Yenthe666** [comment]
@nhomar,

Perhaps a bit of a hard response. It is good that Odoo is adding this in and improving these behaviours.
Nevertheless, @xmo-odoo and @Elkasitu I do think Odoo should look at the password tools for OCA. They have quite a lot of features and it is never bad to look at the 'competition'. 
I believe both OCA and Odoo can learn from eachother. 😉 

Regards,
Yenthe

**Comment by nhomar** [comment]
@Yenthe666 

Never take an answer from myself as **hard**!. strong things must be said directly that's it (none side intentions).

The fact of have a feature like this as official instead a really nice feature already done and at least on our side very well tested.

Friends say things direct and clear! ;-)

I just hope be listen here.

Regards.

**Comment by Yenthe666** [comment]
@nhomar,

Gotcha! 😉 
About the designs and charity: I don't say they should copycat the code blindly or paste it over. Just look at the concepts and ideas behind it, then to the same from the Odoo point of view and combine those two to the best end result (in Odoo's opinion ofcourse).
Anyways I'm happy this PR is here.

Regards,
Yenthe

**Comment by nhomar** [comment]
>> I don't say they should copycat the code blindly or paste it over. Just look at the concepts and ideas behind it, then to the same from the Odoo point of view and combine those two to the best end result (in Odoo's opinion ofcourse).

That's the point.

I know if @xmo-odoo read the code there, it is a feature that should land REALLY cleaner in the **odoo** lib itself instead an external module.

From a technical stand point:

The login process is not clear enough to inherit yet in the partner side, and the behavior when website installed or not is different, and IMHO that investment should be nice to use the a likely python approach as the OCA ones's but fancy with the javascript layer as this one.

I agreed with the point of a copycat is not ok...

File: addons/auth_password_policy/models/res_config_settings.py (hunk 1 above)
Comment (xmo-odoo) [reply addons/auth_password_policy/models/res_config_settings.py]: > what do you think of having a maxlength as well?
 
Lots of bad things.

File: addons/auth_password_policy/static/src/js/password_field.js [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+// Recommendations from Shay (2016)
+// Our research has shown that there are other policies that are more usable
+// and more secure. We found three policies – 2class12, 3class12, and 2word16
+// – that we can directly recommend over comp8
+//
+// Since 2class12 is a superset of 3class12 and 2word16, either pick it or
+// pick the other two (and get the highest score of the two). We're picking
+// the other two.
+var recommandations = [
```
Comment (juan-oa) [inline addons/auth_password_policy/static/src/js/password_field.js]: *recommendations

File: addons/auth_password_policy/static/src/js/password_field.js [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+     * @override
+     * @private
+     */
+    _onInput: function () {
+        this._super();
+
+        // Compute quality factor of password & update meter
+        var password = this._getValue();
+        var actual = this._policy.score(password);
+        // get the highest score of all recommandations
```
Comment (mreficent) [inline addons/auth_password_policy/static/src/js/password_field.js]: recommendations

File: addons/auth_password_policy/static/src/js/password_field.js [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+        var recommended = Math.max.apply(null,
+            recommandations.map(function (p) {
```
Comment (mreficent) [inline addons/auth_password_policy/static/src/js/password_field.js]: recommendations

**Comment by nhomar** [comment]
@odony with that kind of supported answers you can say everything as direct as you want ;-).

Totally understood and from now on I will use your argument (which is so better supported that at the very beginning, because now it is not just a personal opinion but something studied and with better fundaments).

BTW, what I think is that the new policy applicability should be possible to be done with a simple extension instead overwrite too much ugly stuff (because of talk about **average** things instead real use cases is wrong also in my opinion, remember there are the lies the f#!+... lies and the statistics), and our OCA module solve those cases (but it is nice if the inheritance can be cleaner.

From now on, then, I am following your lead.

Regards.

**Comment by hbto** [comment]
@odony @nhomar I have come late to the party, though better late than never.

Is there a way to include MFA/TFA into the party of strengthening  measures
to increase the security into odoo.

It is know fact that user-friendliness and security are not hand-in-hand but
having at least Two Factors to authenticate, a) password (something) you
know b) Gridcard/OTP (something you have) can overwhelmingly increase the
security of Odoo.

OCA has already been developing some modules
https://github.com/OCA/server-tools/tree/10.0/auth_totp

I have worked too on private project on MFA on my own at Vauxoo (kind of
pet projects)

Having users asked about two different things is not the same as having
users remembering strong passwords.

Regards.


Quien suscribe,

*Hbto*

**Comment by gdgellatly** [comment]
@odony @nhomar I'm not sure if I've read Shay.  It seems familiar.  Latest NIST guidelines certainly support it anyway.

I finally tired of seeing support staff spend their Monday mornings with the weekend forgetters every 90 days, so about 3 years ago I transitioned everybody to Oauth, put lastpass into our standard browser build, ran a few training sessions and set 2FA with the oauth providers for employees.  Never had to worry about a password issue since.  Best part is, users would happily give up their "erp" password to anyone who asks, email password not so much.

It is so non intrusive, every morning they login to their email anyway do the 2FA dance and away they go.  With Windows 10 if you are using Azure AD and/or 365 its practically SSO.

The only real user/password feature I need is the ability to turn them off and rely solely on oauth providers.  With browser containers going mainstream its even more feasible.

IMO Odoo should just offer their odoo.com oauth as free (or chargeable if they want) configured/installed by default, robust external oauth/IDP integrations and get rid of user/password entirely or else make it opt in for those 1 in 10,000 cases where external connectivity is not an option or for dev.

**Comment by moylop260** [comment]
Talking about...
The following twitter thread is very useful about the matter: 
 - https://twitter.com/AaronToponce/status/988761412464857088

Norton Signup 😃:
 - screen shot 2018-04-25 at 14 45 27
 - screen shot 2018-04-25 at 14 45 16

**Comment by moylop260** [comment]
FYI in the twitter thread, they are mentioning that Norton will use a package called `zxcvbn` something like as dropbox uses: https://github.com/dropbox/zxcvbn

For python I found the following library: https://pypi.org/project/zxcvbn/
Maybe it could interest you

**Comment by dek-odoo** [comment]
This might be interesting.
https://blog.codinghorror.com/password-rules-are-bullshit/

incorrect-password

:smile:

**Comment by xmo-odoo** [comment]
missing: auth_signup (?) support

**Comment by odony** [comment]
Thanks! Merged at c9a5dc9d0f5f6438cdaf2e96e50c5020375d169c after fixing a BS4 issue in config settings :+1:

**Pull request closed without merging**